server/plugin.go

```diff
@@ -20,9 +23,82 @@ type Plugin struct {
 	configuration *configuration
 }
 
-// ServeHTTP demonstrates a plugin that handles HTTP requests by greeting the world.
+type apiResponse struct {
+	CookieName  string `json:"cookieName"`
+	CookieValue string `json:"cookieValue"`
+	StatusCode  int    `json:"statusCode"`
+}
+
+// ServeHTTP handles HTTP requests.
 func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
-	fmt.Fprint(w, "Hello, world!")
+	switch path := r.URL.Path; path {
+	case "/api/v1/check":
+		p.handleCookieCheck(w, r)
+	}
 }
 
-// See https://developers.mattermost.com/extend/plugins/server/reference/
+// handleCookieCheck function is used to handle the checks for the existense of the Canary cookie.
+func (p *Plugin) handleCookieCheck(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+	if userID == "" {
+		http.Error(w, "Not authorized", http.StatusUnauthorized)
+		return
+	}
+	cookie, err := r.Cookie("canary")
+	if err != nil {
+		p.API.LogDebug("Canary cookie does not exist for user " + userID)
+		err := p.canaryPercentage(w, userID)
+		if err != nil {
+			p.API.LogError("Unable to set cookie" + err.Error())
+			http.Error(w, "Unable to set cookie", http.StatusInternalServerError)
+			return
+		}
+	} else if cookie.Value == "never" {
+		p.API.LogDebug("Canary cookie is set to never for user " + userID)
+		err := p.canaryPercentage(w, userID)
+		if err != nil {
+			p.API.LogError("Unable to set cookie" + err.Error())
+			http.Error(w, "Unable to set cookie", http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+// canaryPercentage function is used to check the canary percentage and call the addCookie function.
+func (p *Plugin) canaryPercentage(w http.ResponseWriter, userID string) error {
+	config := p.getConfiguration()
+	percentage, err := strconv.Atoi(config.CanaryPercentage)
+	if err != nil {
+		return err
+	}
+	randomNumber := rand.Intn(100)
+	if randomNumber < percentage {
+		p.API.LogDebug("Setting Canary cookie to always for user " + userID)
+		p.addCookie(w, "canary", "always")
+	} else {
+		p.API.LogDebug("Setting Canary cookie to never for user " + userID)
+		p.addCookie(w, "canary", "never")
+	}
+	return nil
+}
+
+// addCookie function is used to set the canary cookie.
+func (p *Plugin) addCookie(w http.ResponseWriter, cookieName, cookieValue string) {
+	expire := time.Now().AddDate(0, 0, 1)
+	canaryCookie := http.Cookie{
+		Name:    cookieName,
+		Value:   cookieValue,
+		Expires: expire,
+		Path:    "/",
+	}
+	http.SetCookie(w, &canaryCookie)
+	var response = apiResponse{cookieName, cookieValue, http.StatusOK}
+	writeAPIResponse(w, response)
```

Comment:
Can we actually move this function call up into the `handleCookieCheck` function? I'd like to keep all the parts responsible for reading and writing to the request in a single function so we don't have to dig through a function stack to find out where the response is being written. What I'd suggest doing is just returning from `canaryPercentage` the values that get set and then use those to write the API response at the bottom of `handleCookieCheck`. If no cookie at all gets set (e.g they already had the cookie) then we should just return blank cookie name/value in the API response

Reply:
Function removed and logic added in the handleCookieCheck function. API response for existing canary always cookie added too.